services/workers/src/test/java/org/sagebionetworks/table/worker/MaterializedViewUpdateWorkerIntegrationTest.java

```diff
@@ -242,6 +253,88 @@ public void tesBuildMaterializedViewIndependentOfDependencySourceTable() throws
+		int numberOfFiles = 5;
+		List<Entity> entites = createProjectHierachy(numberOfFiles);
+		EntityView ev = createEntityView(entites);
+
+		Project project = entites.stream().filter(e -> e instanceof Project).map(e -> (Project) e).findFirst().get();
+		Long projectId = KeyFactory.stringToKey(project.getId());
+		// the user can only see files with the project as their benefactor.
+		List<String> fileIdsUserCanSee = entites.stream()
+				.filter((e) -> e instanceof FileEntity
+						&& projectId.equals(entityManager.getEntityHeader(adminUserInfo, e.getId()).getBenefactorId()))
+				.map(e -> e.getId()).collect(Collectors.toList());
+		assertEquals(3, fileIdsUserCanSee.size());
+
+		// Currently do not support doubles so the double key is excluded.
+		String definingSql = "select id, stringKey, longKey, doubleKey, dateKey, booleanKey from " + ev.getId();
+
+		IdAndVersion mvId = createMaterializedView(ev.getParentId(), definingSql);
+
+		String finalSql = "select * from " + mvId + " order by id asc";
+
+		List<Row> expectedRows = Arrays.asList(
+				new Row().setRowId(1L).setVersionNumber(0L)
+						.setValues(Arrays.asList(fileIdsUserCanSee.get(0), "a string: 3", "8", "6.140000000000001",
+								"1004", "false")),
+				new Row().setRowId(3L).setVersionNumber(0L).setValues(
+						Arrays.asList(fileIdsUserCanSee.get(1), "a string: 5", "10", "8.14", "1006", "false")),
+				new Row().setRowId(5L).setVersionNumber(0L).setValues(
+						Arrays.asList(fileIdsUserCanSee.get(2), "a string: 7", "12", "10.14", "1008", "false")));
+		IdAndVersion evId = IdAndVersion.parse(ev.getId());
+
+
+		asyncHelper.assertQueryResult(userInfo, finalSql, (results) -> {
+			assertEquals(expectedRows, results.getQueryResult().getQueryResults().getRows());
+		}, MAX_WAIT_MS);
+
+
+		// create snapshot of underlying entity view for materialized view
+		IdAndVersion evSnapshotId = createSnapshot(evId);
+
+		// wait for table to be available.
+		asyncHelper.assertQueryResult(userInfo, "select * from " + evSnapshotId, (results) -> {
+			assertFalse(results.getQueryResult().getQueryResults().getRows().isEmpty());
+		}, MAX_WAIT_MS);
+
+		//change the defining sql of MV
+		String updatedDefiningSql = "select id, stringKey, longKey, doubleKey, dateKey, booleanKey from " + evSnapshotId;
+		MaterializedView mv = (MaterializedView) entityManager.getEntity(adminUserInfo, mvId.getId().toString());
+		mv.setDefiningSQL(updatedDefiningSql);
+		entityManager.updateEntity( adminUserInfo, mv, false, null);
+
+		// send a message to rebuild the view with updated defining sql
+		ChangeMessage message = new ChangeMessage();
+		message.setChangeType(ChangeType.UPDATE);
+		message.setObjectType(ObjectType.MATERIALIZED_VIEW);
+		message.setObjectId(mv.getId());
+		message = changeDAO.replaceChange(message);
+		repositoryMessagePublisher.publishToTopic(message);
+
+		//call under test
+		TimeUtils.waitFor(MAX_WAIT_MS, 1000L, () -> {
```

Comment:
The first time the query is run it might fail if the MV has not been updated yet.  However, after the MV worker updates the index, this query will eventually pass without error.